src/DDMinimizer.cpp

```diff
@@ -0,0 +1,445 @@
+#include "DDMinimizer.hpp"
+
+#include "Definitions.hpp"
+#include "circuit_optimizer/CircuitOptimizer.hpp"
+#include "ir/Permutation.hpp"
+#include "ir/QuantumComputation.hpp"
+#include "ir/operations/Control.hpp"
+#include "ir/operations/OpType.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <map>
+#include <numeric>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+using namespace qc;
+using namespace std;
+
+namespace qc {
+
+/***
+ * Public Methods
+ ***/
+
+void DDMinimizer::optimizeInputPermutation(qc::QuantumComputation& qc) {
+  // create, set and apply the heuristics based permutation
+  const qc::Permutation perm = DDMinimizer::createGateBasedPermutation(qc);
+  qc.initialLayout = perm;
+  qc::CircuitOptimizer::elidePermutations(qc);
+}
+
+qc::Permutation
+DDMinimizer::createGateBasedPermutation(qc::QuantumComputation& qc) {
+
+  // create the data structure to store the indices in the pattern maps as well
+  // as the max indices of the ladders
+  std::map<std::string, std::map<std::pair<Qubit, Qubit>, int>> maps =
+      DDMinimizer::makeDataStructure(qc).first;
+  std::map<std::string, std::vector<int>> indices =
+      DDMinimizer::makeDataStructure(qc).second;
+  const std::size_t bits = qc.getNqubits();
+
+  // iterate over all the ops and mark the index of the found x-c pairs in the
+  // map.
+  int instructionIndex = 0;
+  bool found = false;
+  for (const auto& op : qc) {
+    if (!op->isStandardOperation() || op->getType() == Z) {
+      continue;
+    }
+    const Controls controls = op->getControls();
+    const std::vector<Qubit> targets = {op->getTargets().begin(),
+                                        op->getTargets().end()};
+
+    for (const auto& control : controls) {
+      for (const auto& target : targets) {
+        for (auto& map : maps) {
+          auto it = map.second.find({control.qubit, target});
+          if (it != map.second.end()) {
+            it->second = instructionIndex;
+            found = true;
+          }
+        }
+      }
+    }
+    instructionIndex++;
+  }
+  if (!found) {
+    return qc.initialLayout;
+  }
+
+  // iterate over all the maps and find the max index of the found x-c pairs
+  for (const auto& map : maps) {
+    int max = -1;
+    for (const auto& pair : map.second) {
+      if (pair.second == -1) {
+        max = -1;
+        break;
+      }
+      if (pair.second > max) {
+        max = pair.second;
+      }
+    }
+    std::string mapName = map.first.substr(0, 2);
+    auto it = indices.find(mapName);
+
+    if (it != indices.end()) {
+      if (mapName[1] == 'R') {
+        const auto column = static_cast<std::size_t>(map.first[2] - '0');
```

Comment:
To substract characters from each other and than convert them to a number seems odd to me. What are the possible values of `map.first[2]`? Are that just a handful different ones that could potentially compared all separately, i.e., with an `if` case for every value.